### PR TITLE
Execute Socket[Channel]#connect under doPrivileged

### PR DESCRIPTION
**Pull request by s1monw**
In order to allow users to run under a security manager that only grants
connect permission to the httpcore codebase the connect methods should
be executed in a doPriveledged block.
This is certainly not the only issue that users run into when they
install a SecurityManager with strict permissions but certainly the
most prominent and most likely one. Upstream components like the client
might also need to protect places accessing the proxy selector etc.

Backport of #135

### CODE REVIEW

**Comment by ok2c** [comment]
@s1monw I tweaked `BasicConnFactory` a bit with 6f72e6c48e8e150d5771a70e7c405e34ce9a08e1. Could you please rebase your PR off the latest 4.4.x?

**Comment by s1monw** [comment]
> Could you please rebase your PR off the latest 4.4.x? 

I can do that. I do wonder why you don't use githubs `squash and merge` from the merge button. I am not sure if it's the default for you but the little drop-down on the merge button gives you the option that way you don't get a merge commit and you don't need to rebase. you can also then edit the commit message?

**Comment by ok2c** [comment]
@s1monw I could. But I would rather see and review the exact change-set to be committed and do `rebase and merge` instead. And I would also like to hope that you spot and eliminate `final Socket finalSocket = socket;` in the process of change-set review.

File: httpcore/src/main/java/org/apache/http/impl/pool/BasicConnFactory.java [hunk 1]
```diff
@@ -177,7 +181,24 @@ public HttpClientConnection create(final HttpHost host) throws IOException {
             socket.setSoLinger(true, linger);
         }
         socket.setKeepAlive(this.sconfig.isSoKeepAlive());
-        socket.connect(new InetSocketAddress(hostname, port), this.connectTimeout);
+        // Run this under a doPrivileged to support lib users that run under a SecurityManager this allows granting connect permissions
+        // only to this library
+        final Socket finalSocket = socket;
```
Comment (ok2c) [inline httpcore/src/main/java/org/apache/http/impl/pool/BasicConnFactory.java]: @s1monw Can we eliminate `finalSocket`? It is no longer required.

Comment (s1monw) [reply httpcore/src/main/java/org/apache/http/impl/pool/BasicConnFactory.java]: done

**Pull request merged**